app/vmalert/rule/group.go

```diff
@@ -436,9 +436,16 @@ func (g *Group) Start(ctx context.Context, nts func() []notifier.Notifier, rw re
 	}
 }
 
-// UpdateWith inserts new group to updateCh
-func (g *Group) UpdateWith(new *Group) {
-	g.updateCh <- new
+// UpdateWithNonBlocking inserts new group to updateCh
+func (g *Group) UpdateWithNonBlocking(new *Group) {
+	select {
+	case g.updateCh <- new:
+	default:
+		// old config hasn't been loaded yet,
+		// drop it and insert the new one.
+		<-g.updateCh
```

Comment:
It's possible race condition, when multiple goroutines updates state.

First goroutine failed to write data into channel. At this time, updateCh value read by group. 

First blocks at `<-g.updateCh`, since there is no value at channel.
Second goroutine pushes update at `case g.updateCh <- new:`, first goroutine reads and discards it.

 As result, update will be lost.  It also produces misleading message, that configuration was updated. But in fact, vmalert uses old configuration.

 Previous behaviour was consistent and less confusing.

Reply:
I removed the lock in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7258/commits/075917a46d7857ea26b7bd9544ac9ff8dba7ec33, because the `UpdateWithNonBlocking()` is only called in hot-reload, and the reload process can't be triggered concurrently. So it's ok for `UpdateWithNonBlocking()` to not be concurrent safe.

>Previous behaviour was consistent and less confusing.

Previously, hot-reload process is blocked when a group is still sleeping in Start(). 
Added more description now, please check.

Reply:
I'm not sure that mutex is good way to go. Maybe atomic.Pointer is better, that holds the latest update. 

 But anyway, it could make update inconsistent.

Reply:
Currently config update executed sequentially for all groups and rules.
UpdateWith called per group and vmalert waits till new configuration updated by waiting at wg.Wait.
It allows vmalert to report that new configuration was applied to the all groups.

By making this call non-blocking, it looks like there is no need to perform wg.Wait operation. Also in-memory state of configuration file could be different compared to actual configuration per group.

Reply:
Ahh, you updated PR. Looks like it should fixed potential config update lock. Looks good!